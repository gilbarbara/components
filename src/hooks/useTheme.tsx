import { useCallback, useRef } from 'react';
import { useTheme as useThemeEmotion } from '@emotion/react';
import deepEqual from '@gilbarbara/deep-equal';

import { mergeTheme } from '~/modules/helpers';

import { Theme } from '~/types';

let storedEmotionTheme: Theme;
let storedTheme: Theme;

export type UseThemeReturn = ReturnType<typeof useTheme>;

export function useTheme() {
  const emotionTheme = useThemeEmotion();
  const needsUpdate = !storedTheme || !deepEqual(storedEmotionTheme, emotionTheme);
  const nextTheme = useRef<Theme>(storedTheme ?? emotionTheme);

  if (needsUpdate) {
    storedEmotionTheme = emotionTheme;
    nextTheme.current = mergeTheme(emotionTheme);
    storedTheme = nextTheme.current;
  }

  const { dataAttributeName } = nextTheme.current;

  const getDataAttributes = useCallback(
    (name: string) => {
      return {
        [`data-${dataAttributeName}`]: name,
      };
    },
    [dataAttributeName],
  );

  return { getDataAttributes, theme: nextTheme.current };
}
