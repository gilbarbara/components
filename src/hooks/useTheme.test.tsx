import { ThemeProvider } from '@emotion/react';
import { renderHook } from '@testing-library/react';

import { useTheme } from './useTheme';

describe('useTheme', () => {
  it('should return properly', () => {
    const { result } = renderHook(() => useTheme());
    const { getDataAttributes, theme } = result.current;

    expect(getDataAttributes('Button')).toEqual({
      'data-component-name': 'Button',
    });
    expect(theme).toMatchSnapshot();
  });

  it('should return a custom theme', () => {
    const customTheme = {
      dataAttributeName: 'testid',
      darkMode: true,
    };
    const { result } = renderHook(() => useTheme(), {
      wrapper: ({ children }) => <ThemeProvider theme={customTheme}>{children}</ThemeProvider>,
    });
    const { getDataAttributes, theme } = result.current;

    expect(getDataAttributes('Button')).toEqual({
      'data-testid': 'Button',
    });
    expect(theme.darkMode).toBeTrue();
  });
});
