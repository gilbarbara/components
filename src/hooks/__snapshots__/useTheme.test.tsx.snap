// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useTheme > should return properly 1`] = `
{
  "avatar": {
    "jumbo": {
      "fontSize": "52px",
      "size": "128px",
    },
    "lg": {
      "fontSize": "24px",
      "size": "64px",
    },
    "md": {
      "fontSize": "18px",
      "size": "48px",
    },
    "sm": {
      "fontSize": "14px",
      "size": "32px",
    },
    "xl": {
      "fontSize": "40px",
      "size": "96px",
    },
    "xs": {
      "fontSize": "12px",
      "size": "24px",
    },
  },
  "black": "#000",
  "breakpoints": {
    "lg": "1024px",
    "md": "768px",
    "sm": "400px",
    "xl": "1280px",
    "xs": "375px",
    "xxl": "1440px",
    "xxs": "320px",
    "xxxl": "1680px",
  },
  "button": {
    "lg": {
      "borderRadius": "8px",
      "fontSize": "18px",
      "fontWeight": 700,
      "height": "48px",
      "lineHeight": 1.4,
      "padding": [
        "10px",
        "20px",
      ],
    },
    "md": {
      "borderRadius": "8px",
      "fontSize": "16px",
      "fontWeight": 700,
      "height": "40px",
      "lineHeight": 1.3,
      "padding": [
        "8px",
        "16px",
      ],
    },
    "sm": {
      "borderRadius": "8px",
      "fontSize": "14px",
      "fontWeight": 700,
      "height": "32px",
      "lineHeight": 1.2,
      "padding": [
        "6px",
        "12px",
      ],
    },
    "xl": {
      "borderRadius": "8px",
      "fontSize": "24px",
      "fontWeight": 700,
      "height": "56px",
      "lineHeight": 1.4,
      "padding": [
        "12px",
        "24px",
      ],
    },
    "xs": {
      "borderRadius": "4px",
      "fontSize": "12px",
      "fontWeight": 700,
      "height": "24px",
      "lineHeight": 1.2,
      "padding": [
        "4px",
        "8px",
      ],
    },
  },
  "colors": {
    "blue": "#66a5ff",
    "citrus": "#a8ff66",
    "cyan": "#0bc5ea",
    "gray": "#b3b3b3",
    "green": "#58d063",
    "indigo": "#8a8fff",
    "lime": "#d9ff66",
    "orange": "#ff995d",
    "pink": "#ee63ab",
    "primary": "#0049DB",
    "purple": "#ad7bff",
    "red": "#ff5e5e",
    "secondary": "#006157",
    "teal": "#52d6c5",
    "yellow": "#ffe166",
  },
  "darkColor": "#262626",
  "darkMode": false,
  "dataAttributeName": "component-name",
  "easing": "cubic-bezier(0.65, 0.815, 0.735, 0.395)",
  "fontFamily": "inherit",
  "fontMonospace": "Courier, monospace",
  "fontWeights": {
    "bold": 700,
    "normal": 400,
  },
  "grayScale": {
    "10": "#fcfcfc",
    "100": "#e6e6e6",
    "150": "#d9d9d9",
    "20": "#fafafa",
    "200": "#cccccc",
    "30": "#f7f7f7",
    "300": "#b3b3b3",
    "40": "#f5f5f5",
    "400": "#999999",
    "50": "#f2f2f2",
    "500": "#808080",
    "600": "#666666",
    "700": "#4d4d4d",
    "750": "#404040",
    "800": "#333333",
    "850": "#262626",
    "900": "#1a1a1a",
    "950": "#0d0d0d",
  },
  "image": {
    "animationDuration": 1.5,
    "radius": "none",
    "skeletonBg": "#f2f2f2",
    "skeletonColor": "#e6e6e6",
    "timingFunction": "ease-in-out",
  },
  "inputHeight": {
    "lg": "48px",
    "md": "40px",
    "sm": "32px",
  },
  "inputPaddingY": {
    "lg": "16px",
    "md": "12px",
    "sm": "8px",
  },
  "lightColor": "#f5f5f5",
  "opacity": {
    "intense": 0.64,
    "light": 0.16,
    "medium": 0.32,
    "semiOpaque": 0.72,
    "semiTransparent": 0.08,
  },
  "opacityDisabled": 0.64,
  "outlineOffset": 1,
  "outlineOpacity": 0.6,
  "outlineWidth": 3,
  "outlineZIndex": 10,
  "radius": {
    "full": "9999px",
    "lg": "24px",
    "md": "16px",
    "none": "0px",
    "round": "50%",
    "sm": "12px",
    "xl": "32px",
    "xs": "8px",
    "xxs": "4px",
  },
  "selectPaddingY": {
    "lg": "12px",
    "md": "8px",
    "sm": "4px",
  },
  "shadow": {
    "high": "2px 4px 16px 8px rgba(148, 148, 148, 0.32)",
    "low": "2px 4px 16px 2px rgba(148, 148, 148, 0.16)",
    "mid": "2px 4px 16px 4px rgba(148, 148, 148, 0.24)",
  },
  "shadowDark": {
    "high": "2px 4px 16px 8px rgba(32, 32, 32, 0.32)",
    "low": "2px 4px 16px 2px rgba(32, 32, 32, 0.16)",
    "mid": "2px 4px 16px 4px rgba(32, 32, 32, 0.24)",
  },
  "spacing": {
    "jumbo": "96px",
    "lg": "24px",
    "md": "16px",
    "sm": "12px",
    "xl": "32px",
    "xs": "8px",
    "xxl": "48px",
    "xxs": "4px",
    "xxxl": "64px",
    "xxxs": "2px",
  },
  "swatchOptions": {},
  "textColorOptions": {
    "threshold": 128,
  },
  "typography": {
    "h1": {
      "fontSize": "32px",
      "lineHeight": 1.2,
    },
    "h2": {
      "fontSize": "28px",
      "lineHeight": 1.2,
    },
    "h3": {
      "fontSize": "24px",
      "lineHeight": 1.2,
    },
    "h4": {
      "fontSize": "20px",
      "lineHeight": 1.2,
    },
    "h5": {
      "fontSize": "18px",
      "lineHeight": 1.2,
    },
    "h6": {
      "fontSize": "16px",
      "lineHeight": 1.2,
    },
    "jumbo": {
      "fontSize": "48px",
      "lineHeight": 1.2,
    },
    "jumboLarge": {
      "fontSize": "64px",
      "lineHeight": 1.2,
    },
    "lg": {
      "fontSize": "18px",
      "lineHeight": 1.2,
    },
    "md": {
      "fontSize": "16px",
      "lineHeight": 1.2,
    },
    "sm": {
      "fontSize": "14px",
      "lineHeight": 1.2,
    },
    "xl": {
      "fontSize": "20px",
      "lineHeight": 1.2,
    },
    "xs": {
      "fontSize": "12px",
      "lineHeight": 1.2,
    },
    "xxs": {
      "fontSize": "10px",
      "lineHeight": 1.1,
    },
  },
  "variants": {
    "blue": {
      "100": "#e1eeff",
      "200": "#c9e0ff",
      "300": "#acceff",
      "400": "#8cb8f7",
      "50": "#eef6ff",
      "500": "#6b9fea",
      "600": "#4a84d6",
      "700": "#2a67bb",
      "800": "#094999",
      "900": "#002c6f",
      "950": "#00113f",
    },
    "citrus": {
      "100": "#e2f4d7",
      "200": "#cbeab9",
      "300": "#afdb93",
      "400": "#90c968",
      "50": "#eff9e9",
      "500": "#6fb435",
      "600": "#4d9a00",
      "700": "#2c7e00",
      "800": "#086000",
      "900": "#004000",
      "950": "#001f00",
    },
    "cyan": {
      "100": "#d9f2fa",
      "200": "#bce7f4",
      "300": "#96d8eb",
      "400": "#69c5de",
      "50": "#eaf8fc",
      "500": "#2eaecd",
      "600": "#0095b7",
      "700": "#00799c",
      "800": "#005b7c",
      "900": "#003c57",
      "950": "#001d2f",
    },
    "gray": {
      "100": "#ededed",
      "200": "#dedede",
      "300": "#cbcbcb",
      "400": "#b6b6b6",
      "50": "#f5f5f5",
      "500": "#9e9e9e",
      "600": "#848484",
      "700": "#686868",
      "800": "#4d4d4d",
      "900": "#313131",
      "950": "#161616",
    },
    "green": {
      "100": "#def4de",
      "200": "#c4ebc4",
      "300": "#a4dda5",
      "400": "#7fcc82",
      "50": "#edf9ed",
      "500": "#56b65d",
      "600": "#259e38",
      "700": "#008208",
      "800": "#006300",
      "900": "#004200",
      "950": "#002100",
    },
    "indigo": {
      "100": "#e8ebff",
      "200": "#d6dbff",
      "300": "#bfc6ff",
      "400": "#a7aefd",
      "50": "#f2f4ff",
      "500": "#8d93f1",
      "600": "#7376dd",
      "700": "#5958c2",
      "800": "#403b9f",
      "900": "#281f74",
      "950": "#110743",
    },
    "lime": {
      "100": "#e9f1d6",
      "200": "#d7e6b6",
      "300": "#c1d68f",
      "400": "#a9c262",
      "50": "#f3f8e8",
      "500": "#8fac2a",
      "600": "#749200",
      "700": "#597600",
      "800": "#3f5800",
      "900": "#263900",
      "950": "#0f1b00",
    },
    "orange": {
      "100": "#fee8dc",
      "200": "#fcd5c1",
      "300": "#f5bea0",
      "400": "#eaa37b",
      "50": "#fff2eb",
      "500": "#da8655",
      "600": "#c3672d",
      "700": "#a74900",
      "800": "#852c00",
      "900": "#5e1100",
      "950": "#330000",
    },
    "pink": {
      "100": "#ffe3f0",
      "200": "#ffcde4",
      "300": "#feb1d4",
      "400": "#f492c0",
      "50": "#ffeff7",
      "500": "#e470aa",
      "600": "#ce4d91",
      "700": "#b12776",
      "800": "#8d0059",
      "900": "#64003b",
      "950": "#37001c",
    },
    "primary": {
      "100": "#dceeff",
      "200": "#c1dfff",
      "300": "#a1cbff",
      "400": "#7db3ff",
      "50": "#ecf6ff",
      "500": "#5997ff",
      "600": "#3578ff",
      "700": "#1157e9",
      "800": "#0036c3",
      "900": "#001492",
      "950": "#000057",
    },
    "purple": {
      "100": "#f0e8ff",
      "200": "#e4d5ff",
      "300": "#d4beff",
      "400": "#c1a2ff",
      "50": "#f7f2ff",
      "500": "#ac84f2",
      "600": "#9465df",
      "700": "#7a45c4",
      "800": "#5e25a1",
      "900": "#3f0475",
      "950": "#200044",
    },
    "red": {
      "100": "#ffe3e0",
      "200": "#ffcdc8",
      "300": "#ffb1ab",
      "400": "#ff918b",
      "50": "#fff0ee",
      "500": "#f36d6a",
      "600": "#dd474a",
      "700": "#c01a2b",
      "800": "#9b000c",
      "900": "#6f0000",
      "950": "#3e0000",
    },
    "secondary": {
      "100": "#e2f1ee",
      "200": "#cbe5e0",
      "300": "#afd5ce",
      "400": "#8fc2b9",
      "50": "#eff7f6",
      "500": "#6daca1",
      "600": "#4a9388",
      "700": "#27776d",
      "800": "#005a51",
      "900": "#003c34",
      "950": "#001d18",
    },
    "teal": {
      "100": "#dcf3ef",
      "200": "#c1e8e1",
      "300": "#9ed9d0",
      "400": "#76c7bb",
      "50": "#ecf8f6",
      "500": "#48b2a4",
      "600": "#00998a",
      "700": "#007d6f",
      "800": "#005f53",
      "900": "#004036",
      "950": "#002019",
    },
    "yellow": {
      "100": "#f3edd7",
      "200": "#e8dfb8",
      "300": "#dacc93",
      "400": "#c9b668",
      "50": "#f8f5e9",
      "500": "#b59d37",
      "600": "#9d8200",
      "700": "#826600",
      "800": "#654900",
      "900": "#442d00",
      "950": "#231300",
    },
  },
  "white": "#fff",
  "zIndex": {
    "lg": 200,
    "md": 100,
    "sm": 50,
    "xl": 500,
    "xs": 20,
    "xxl": 1000,
    "xxs": 10,
  },
}
`;
